src/include/distributed/metadata_utility.h

```diff
@@ -203,6 +203,71 @@ typedef enum SizeQueryType
 	TABLE_SIZE /* pg_table_size() */
 } SizeQueryType;
 
+typedef enum BackgroundJobStatus
+{
+	BACKGROUND_JOB_STATUS_SCHEDULED,
+	BACKGROUND_JOB_STATUS_RUNNING,
+	BACKGROUND_JOB_STATUS_FINISHED,
+	BACKGROUND_JOB_STATUS_CANCELLED,
+	BACKGROUND_JOB_STATUS_FAILING,
+	BACKGROUND_JOB_STATUS_FAILED
+} BackgroundJobStatus;
+
+typedef struct BackgroundJob
+{
+	int64 jobid;
+	BackgroundJobStatus state;
+	char *jobType;
+	char *description;
+	TimestampTz *started_at;
+	TimestampTz *finished_at;
+
+	/* extra space to store values for nullable value types above */
+	struct
+	{
+		TimestampTz started_at;
+		TimestampTz finished_at;
+	} __do_not_use;
+} BackgroundJob;
+
+typedef enum BackgroundTaskStatus
+{
+	BACKGROUND_TASK_STATUS_BLOCKED,
+	BACKGROUND_TASK_STATUS_RUNNABLE,
+	BACKGROUND_TASK_STATUS_RUNNING,
+	BACKGROUND_TASK_STATUS_DONE,
+	BACKGROUND_TASK_STATUS_ERROR,
+	BACKGROUND_TASK_STATUS_UNSCHEDULED,
+	BACKGROUND_TASK_STATUS_CANCELLED
+} BackgroundTaskStatus;
+
+typedef struct BackgroundTask
+{
+	int64 jobid;
+	int64 taskid;
+	Oid owner;
+	int32 *pid;
+	BackgroundTaskStatus status;
+	char *command;
+	int32 *retry_count;
+	TimestampTz *not_before;
+	char *message;
+
+	/* extra space to store values for nullable value types above */
+	struct
+	{
+		int32 pid;
+		int32 retry_count;
+		TimestampTz not_before;
+	} __do_not_use;
+} BackgroundTask;
+
+#define SET_NULLABLE_FIELD(ptr, field, value) \
+	(ptr)->__do_not_use.field = (value); \
+	(ptr)->field = &((ptr)->__do_not_use.field)
```

Comment:
the use of the `__do_not_use` property seems a quite scary. Is there a way to achieve the same without using it? If not, this at least needs a comment explaining why it's necessary.

Reply:
I tried `__internal` first, but my compiler complained.

There are some values, like retry_count, where `0` could have a semantically different meaning compared to a `NULL` value.

At first I was allocating small memory chunks to store the values in, which became pretty cumbersome.

Only other way I could think about is adding separate bool values indicating that the field was NULL and the value should be ignored. If you think adding bool values for the NULLABLE value fields I could refactor it that way.

There are still `char *`, but those are separately allocated already, and that seems to be more common anyway, compared to `TimestampTz` and `int32` fields.

Reply:
Ah, after reviewing more I now realize that we create `__do_not_use` ourselves. I thought it was a postgres thing.

Reply:
I could probably rename this to `__extra_storage` or something, to make it less scary sounding.